python/dgl/heterograph.py

```diff
@@ -0,0 +1,2092 @@
+        query with
+
+        .. code::
+
+           vid in g['vtype']
+
+        Examples
+        --------
+        >>> 0 in g['user']
+        True
+        """
+        pass
+
+    def has_nodes(self, vids):
+        """Return a 0-1 array ``a`` given the node ID array ``vids``.
+
+        ``a[i]`` is 1 if the graph contains node ``vids[i]``, 0 otherwise.
+
+        Only works if the graph has one node type.  For multiple types,
+        query with
+
+        .. code::
+
+           g['vtype'].has_nodes(vids)
+
+        Parameters
+        ----------
+        vid : list or tensor
+            The array of node IDs.
+
+        Returns
+        -------
+        a : tensor
+            0-1 array indicating existence
+
+        Examples
+        --------
+        The following example uses PyTorch backend.
+
+        >>> g['user'].has_nodes([0, 1, 2, 3, 4])
+        tensor([1, 1, 1, 0, 0])
+
+        See Also
+        --------
+        has_node
+        """
+        pass
+
+    def has_edge_between(self, u, v):
+        """Return True if the edge (u, v) is in the graph.
+
+        Only works if the graph has one edge type.  For multiple types,
+        query with
+
+        .. code::
+
+           g['srctype', 'dsttype', 'edgetype'].has_edge_between(u, v)
+
+        Parameters
+        ----------
+        u : int
+            The node ID of source type.
+        v : int
+            The node ID of destination type.
+
+        Returns
+        -------
+        bool
+            True if the edge is in the graph, False otherwise.
+
+        Examples
+        --------
+        Check whether Alice plays Tetris
+        >>> g['user', 'game', 'plays'].has_edge_between(0, 1)
+        True
+
+        And whether Alice plays Minecraft
+        >>> g['user', 'game', 'plays'].has_edge_between(0, 2)
+        False
+
+        See Also
+        --------
+        has_edges_between
+        """
+        pass
+
+    def has_edges_between(self, u, v):
+        """Return a 0-1 array `a` given the source node ID array `u` and
+        destination node ID array `v`.
+
+        `a[i]` is 1 if the graph contains edge `(u[i], v[i])`, 0 otherwise.
+
+        Only works if the graph has one edge type.  For multiple types,
+        query with
+
+        .. code::
+
+           g['srctype', 'dsttype', 'edgetype'].has_edges_between(u, v)
+
+        Parameters
+        ----------
+        u : list, tensor
+            The node ID array of source type.
+        v : list, tensor
+            The node ID array of destination type.
+
+        Returns
+        -------
+        a : tensor
+            0-1 array indicating existence.
+
+        Examples
+        --------
+        The following example uses PyTorch backend.
+
+        >>> g['user', 'game', 'plays'].has_edges_between([0, 0], [1, 2])
+        tensor([1, 0])
+
+        See Also
+        --------
+        has_edge_between
+        """
+        pass
+
+    def predecessors(self, v):
```

Comment:
this method needs node type. otherwise, it doesn't work with bipartite graph.

Reply:
If we require the subgraph to have only one edge type then it does.  `v` would always be of the node type of destination.

Reply:
so you assume that all edges in a graph go from one type of nodes to another type? Is it always true?

Reply:
We assume that all edges belonging to the same edge type always have the same source node type and destination node type (which is the assumption of the metagraph)